File: src/main/java/com/itsz/portfolio/service/calculator/CallOptionPriceCalculator.java
package com.itsz.portfolio.service.calculator;

import com.itsz.portfolio.entity.Security;
import com.itsz.portfolio.entity.SecurityType;
import org.springframework.stereotype.Component;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;

@Component
final class CallOptionPriceCalculator implements SecurityPriceCalculator {

    @Override
    public BigDecimal calculate(Security security, BigDecimal price) {
        int years = security.getMaturity().getYear() - LocalDate.now().getYear();
        return BigDecimal.valueOf(BlackScholesUtil.calculateCallOptionPrice(price.doubleValue(), security.getStrike().doubleValue(), years)).setScale(2, RoundingMode.HALF_UP);
    }

    @Override
    public SecurityType getSecurityType() {
        return SecurityType.CALL;
    }


}

File: src/main/kotlin/com/itsz/portfolio/service/calculator/CallOptionPriceCalculator.kt
package com.itsz.portfolio.service.calculator

import com.itsz.portfolio.entity.Security
import com.itsz.portfolio.entity.SecurityType
import org.springframework.stereotype.Component
import java.math.BigDecimal
import java.math.RoundingMode
import java.time.LocalDate

@Component
internal class CallOptionPriceCalculator : SecurityPriceCalculator {

    override fun calculate(security: Security, price: BigDecimal): BigDecimal {
        val years = (security.maturity!!.year - LocalDate.now().year).toDouble()
        return BigDecimal.valueOf(
            BlackScholesUtil.calculateCallOptionPrice(
                price.toDouble(), security.strike!!.toDouble(), years
            )
        ).setScale(2, RoundingMode.HALF_UP)
    }

    override fun getSecurityType(): SecurityType = SecurityType.CALL
}
